File: backend/app/services/crawler/crawler.py
"""Crawler: fetches a small, bounded set of pages for analysis.

MVP limits (configurable via env):
  * MAX_PAGES pages (default 5)
  * depth 1 (links found on the start page only)
  * per-request timeout and byte cap handled by fetcher
"""
from __future__ import annotations

from concurrent.futures import ThreadPoolExecutor, as_completed
from dataclasses import dataclass, field
from typing import Callable, Dict, List, Optional
from urllib.parse import urljoin, urlparse, urlunparse

import httpx

from app.core.config import settings
from app.services.crawler.fetcher import FetchError, FetchedPage, fetch_page
from app.services.crawler.html_parser import ParsedPage, parse_page
from app.services.crawler.url_validator import get_domain, validate_url


@dataclass
class RobotsInfo:
    available: bool = False
    status_code: Optional[int] = None
    disallow_all: bool = False
    disallowed_paths: List[str] = field(default_factory=list)
    sitemap_urls: List[str] = field(default_factory=list)


@dataclass
class SitemapInfo:
    available: bool = False
    url: Optional[str] = None
    status_code: Optional[int] = None


@dataclass
class CrawlContext:
    start_url: str
    domain: str
    pages: List[FetchedPage] = field(default_factory=list)
    parsed_pages: List[ParsedPage] = field(default_factory=list)
    robots: RobotsInfo = field(default_factory=RobotsInfo)
    sitemap: SitemapInfo = field(default_factory=SitemapInfo)
    crawl_limited: bool = False
    broken_links: List[Dict] = field(default_factory=list)
    fetch_errors: List[Dict] = field(default_factory=list)

    @property
    def start_page(self) -> Optional[FetchedPage]:
        return self.pages[0] if self.pages else None

    @property
    def start_parsed(self) -> Optional[ParsedPage]:
        return self.parsed_pages[0] if self.parsed_pages else None


ProgressCallback = Callable[[int, str], None]


def _noop(progress: int, stage: str) -> None:
    pass


def _parse_robots(text: str, base_url: str) -> RobotsInfo:
    info = RobotsInfo(available=True)
    current_agents: List[str] = []
    global_disallow = False
    for raw_line in text.splitlines():
        line = raw_line.split("#", 1)[0].strip()
        if not line:
            continue
        if ":" not in line:
            continue
        key, value = line.split(":", 1)
        key = key.strip().lower()
        value = value.strip()
        if key == "user-agent":
            current_agents = [value.lower()]
        elif key == "disallow" and value and ("*" in current_agents or "aiwebsiteauditor" in current_agents):
            if value == "/":
                global_disallow = True
            info.disallowed_paths.append(value)
        elif key == "sitemap" and value:
            info.sitemap_urls.append(urljoin(base_url, value))
    info.disallow_all = global_disallow
    return info


def _probe_url(url: str) -> Optional[int]:
    """Lightweight availability probe.

    Redirects are intentionally NOT followed automatically: every hop in the
    main fetcher is re-validated against SSRF rules, and probes must not open
    unvalidated destinations. A 3xx status is treated as available.
    """
    try:
        validated = validate_url(url)
    except Exception:
        return None
    with httpx.Client(
        timeout=min(settings.REQUEST_TIMEOUT_SECONDS, 10.0),
        follow_redirects=False,
        headers={"User-Agent": settings.USER_AGENT},
    ) as client:
        try:
            resp = client.head(validated.url)
            if resp.status_code in (405, 501):
                resp = client.get(validated.url, stream=True)
            return resp.status_code
        except httpx.HTTPError:
            return None


def _extract_internal_links(parsed: ParsedPage, domain: str) -> List[str]:
    seen = set()
    out = []
    for link in parsed.links:
        href = link["absolute"]
        if not href.startswith(("http://", "https://")):
            continue
        host = (urlparse(href).hostname or "").lower()
        if host == domain or host == f"www.{domain}" or f"www.{host}" == domain:
            # Drop fragments and query-only duplicates.
            clean = href.split("#", 1)[0]
            if clean not in seen:
                seen.add(clean)
                out.append(clean)
    return out


def crawl(
    raw_url: str,
    progress: ProgressCallback = _noop,
    max_pages: Optional[int] = None,
) -> CrawlContext:
    """Crawl a site with strict limits and return everything the analyzers need."""
    validated = validate_url(raw_url)
    domain = get_domain(validated.url)
    ctx = CrawlContext(start_url=validated.url, domain=domain)
    page_limit = max_pages or settings.MAX_PAGES

    # 1. Start page
    progress(8, "Fetching homepage")
    start = fetch_page(validated.url)
    if start.status_code >= 500:
        raise FetchError(
            f"Start page returned {start.status_code}",
            f"The website returned a server error (HTTP {start.status_code}).",
            kind="http_5xx",
        )
    ctx.pages.append(start)
    ctx.parsed_pages.append(parse_page(start))

    # 2. robots.txt
    progress(16, "Checking robots.txt")
    robots_url = urlunparse((validated.scheme, f"{domain}", "/robots.txt", "", "", ""))
    try:
        robots_resp = fetch_page(robots_url)
        ctx.robots.status_code = robots_resp.status_code
        if robots_resp.status_code < 400 and robots_resp.text:
            parsed_robots = _parse_robots(robots_resp.text, robots_resp.final_url)
            parsed_robots.status_code = robots_resp.status_code
            ctx.robots = parsed_robots
    except FetchError:
        pass  # robots.txt unavailable is a finding, not a crash

    # 3. sitemap
    progress(24, "Checking sitemap")
    sitemap_candidates = list(ctx.robots.sitemap_urls[:2])
    sitemap_candidates.append(urlunparse((validated.scheme, f"{domain}", "/sitemap.xml", "", "", "")))
    for candidate in sitemap_candidates:
        status = _probe_url(candidate)
        if status and status < 400:
            ctx.sitemap = SitemapInfo(available=True, url=candidate, status_code=status)
            break
        elif status:
            ctx.sitemap = SitemapInfo(available=False, url=candidate, status_code=status)

    # 4. Internal pages (depth 1)
    progress(32, "Discovering pages")
    start_parsed = ctx.start_parsed
    internal_links = _extract_internal_links(start_parsed, domain) if start_parsed else []
    max_extra = max(0, page_limit - 1)
    if max_extra > 0 and internal_links:
        to_fetch = internal_links[:max_extra]
        with ThreadPoolExecutor(max_workers=min(settings.CRAWLER_MAX_CONCURRENCY, len(to_fetch))) as pool:
            futures = {pool.submit(fetch_page, url): url for url in to_fetch}
            done = 0
            for fut in as_completed(futures):
                done += 1
                progress(32 + int(24 * done / max(1, len(futures))), f"Crawling page {done} of {len(futures)}")
                url = futures[fut]
                try:
                    page = fut.result()
                    if page.status_code < 400:
                        ctx.pages.append(page)
                        ctx.parsed_pages.append(parse_page(page))
                    else:
                        ctx.fetch_errors.append(
                            {"url": url, "kind": "http_status", "status_code": page.status_code, "message": f"HTTP {page.status_code}"}
                        )
                except FetchError as exc:
                    ctx.fetch_errors.append(
                        {"url": url, "kind": exc.kind, "message": exc.safe_message}
                    )
        if len(internal_links) > max_extra:
            ctx.crawl_limited = True

    # 5. Broken internal links (practical, bounded)
    progress(60, "Checking links")
    if start_parsed:
        candidates = [u for u in internal_links[:12] if u not in {p.url for p in ctx.pages}]
        for url in candidates:
            status = _probe_url(url)
            if status and status >= 400:
                ctx.broken_links.append({"url": url, "status_code": status})

    progress(66, "Analyzing pages")
    return ctx

File: backend/src/services/crawler/crawler.ts
import { config } from "../../config.js";
import { FetchError, fetchPage, probeUrl, type FetchedPage } from "./fetcher.js";
import { parsePage, type ParsedPage } from "./htmlParser.js";
import { getDomain, validateUrl } from "./urlValidator.js";

export interface RobotsInfo {
  available: boolean;
  statusCode: number | null;
  disallowAll: boolean;
  disallowedPaths: string[];
  sitemapUrls: string[];
}

export interface SitemapInfo {
  available: boolean;
  url: string | null;
  statusCode: number | null;
}

export interface CrawlContext {
  startUrl: string;
  domain: string;
  pages: FetchedPage[];
  parsedPages: ParsedPage[];
  robots: RobotsInfo;
  sitemap: SitemapInfo;
  crawlLimited: boolean;
  brokenLinks: { url: string; status_code: number }[];
  fetchErrors: { url: string; kind: string; message: string; status_code?: number }[];
}

export type ProgressCallback = (progress: number, stage: string) => void;

function noop(_progress: number, _stage: string): void {}

function parseRobots(text: string, baseUrl: string): RobotsInfo {
  const info: RobotsInfo = {
    available: true,
    statusCode: null,
    disallowAll: false,
    disallowedPaths: [],
    sitemapUrls: [],
  };
  let currentAgents: string[] = [];
  let globalDisallow = false;

  for (const rawLine of text.split("\n")) {
    const line = rawLine.split("#", 1)[0].trim();
    if (!line || !line.includes(":")) continue;
    const colonIndex = line.indexOf(":");
    const key = line.slice(0, colonIndex).trim().toLowerCase();
    const value = line.slice(colonIndex + 1).trim();
    if (key === "user-agent") {
      currentAgents = [value.toLowerCase()];
    } else if (key === "disallow" && value) {
      if (currentAgents.includes("*") || currentAgents.includes("aiwebsiteauditor")) {
        if (value === "/") globalDisallow = true;
        info.disallowedPaths.push(value);
      }
    } else if (key === "sitemap" && value) {
      try {
        info.sitemapUrls.push(new URL(value, baseUrl).toString());
      } catch {
        // ignore invalid sitemap URL
      }
    }
  }
  info.disallowAll = globalDisallow;
  return info;
}

function extractInternalLinks(parsed: ParsedPage, domain: string): string[] {
  const seen = new Set<string>();
  const out: string[] = [];
  for (const link of parsed.links) {
    const href = link.absolute;
    if (!href.startsWith("http://") && !href.startsWith("https://")) continue;
    let host = "";
    try {
      host = new URL(href).hostname.toLowerCase();
    } catch {
      continue;
    }
    if (host === domain || host === `www.${domain}` || `www.${host}` === domain) {
      const cleanHref = href.split("#", 1)[0];
      if (!seen.has(cleanHref)) {
        seen.add(cleanHref);
        out.push(cleanHref);
      }
    }
  }
  return out;
}

async function mapConcurrent<T, R>(
  items: T[],
  concurrency: number,
  fn: (item: T) => Promise<R>
): Promise<R[]> {
  const results: R[] = new Array(items.length);
  let nextIndex = 0;
  const workers = Array.from({ length: Math.min(concurrency, items.length) }, async () => {
    for (;;) {
      const index = nextIndex;
      nextIndex += 1;
      if (index >= items.length) return;
      results[index] = await fn(items[index]);
    }
  });
  await Promise.all(workers);
  return results;
}

export async function crawl(
  rawUrl: string,
  progress: ProgressCallback = noop,
  maxPages?: number
): Promise<CrawlContext> {
  const validated = await validateUrl(rawUrl);
  const domain = getDomain(validated.url);
  const ctx: CrawlContext = {
    startUrl: validated.url,
    domain,
    pages: [],
    parsedPages: [],
    robots: {
      available: false,
      statusCode: null,
      disallowAll: false,
      disallowedPaths: [],
      sitemapUrls: [],
    },
    sitemap: { available: false, url: null, statusCode: null },
    crawlLimited: false,
    brokenLinks: [],
    fetchErrors: [],
  };
  const pageLimit = maxPages || config.maxPages;

  // 1. Start page
  progress(8, "Fetching homepage");
  const start = await fetchPage(validated.url);
  if (start.statusCode >= 500) {
    throw new FetchError(
      `Start page returned ${start.statusCode}`,
      `The website returned a server error (HTTP ${start.statusCode}).`,
      "http_5xx"
    );
  }
  ctx.pages.push(start);
  ctx.parsedPages.push(parsePage(start));

  // 2. robots.txt
  progress(16, "Checking robots.txt");
  const robotsUrl = `${validated.scheme}://${domain}/robots.txt`;
  try {
    const robotsResp = await fetchPage(robotsUrl);
    ctx.robots.statusCode = robotsResp.statusCode;
    if (robotsResp.statusCode < 400 && robotsResp.content.length > 0) {
      const parsedRobots = parseRobots(decodeText(robotsResp), robotsResp.finalUrl);
      parsedRobots.statusCode = robotsResp.statusCode;
      ctx.robots = parsedRobots;
    }
  } catch {
    // robots.txt unavailable is a finding, not a crash
  }

  // 3. sitemap
  progress(24, "Checking sitemap");
  const sitemapCandidates = [...ctx.robots.sitemapUrls.slice(0, 2), `${validated.scheme}://${domain}/sitemap.xml`];
  for (const candidate of sitemapCandidates) {
    const status = await probeUrl(candidate);
    if (status !== null && status < 400) {
      ctx.sitemap = { available: true, url: candidate, statusCode: status };
      break;
    } else if (status !== null) {
      ctx.sitemap = { available: false, url: candidate, statusCode: status };
    }
  }

  // 4. Internal pages (depth 1)
  progress(32, "Discovering pages");
  const startParsed = ctx.parsedPages[0];
  const internalLinks = startParsed ? extractInternalLinks(startParsed, domain) : [];
  const maxExtra = Math.max(0, pageLimit - 1);
  if (maxExtra > 0 && internalLinks.length > 0) {
    const toFetch = internalLinks.slice(0, maxExtra);
    let done = 0;
    await mapConcurrent(
      toFetch,
      Math.min(config.crawlerMaxConcurrency, toFetch.length),
      async (url) => {
        done += 1;
        progress(32 + Math.floor((24 * done) / Math.max(1, toFetch.length)), `Crawling page ${done} of ${toFetch.length}`);
        try {
          const page = await fetchPage(url);
          if (page.statusCode < 400) {
            ctx.pages.push(page);
            ctx.parsedPages.push(parsePage(page));
          } else {
            ctx.fetchErrors.push({
              url,
              kind: "http_status",
              status_code: page.statusCode,
              message: `HTTP ${page.statusCode}`,
            });
          }
        } catch (err) {
          if (err instanceof FetchError) {
            ctx.fetchErrors.push({ url, kind: err.kind, message: err.safeMessage });
          } else {
            ctx.fetchErrors.push({ url, kind: "fetch_error", message: "Failed to fetch page." });
          }
        }
      }
    );
    if (internalLinks.length > maxExtra) ctx.crawlLimited = true;
  }

  // 5. Broken internal links (practical, bounded)
  progress(60, "Checking links");
  if (startParsed) {
    const fetchedUrls = new Set(ctx.pages.map((p) => p.url));
    const candidates = internalLinks.slice(0, 12).filter((u) => !fetchedUrls.has(u));
    for (const url of candidates) {
      const status = await probeUrl(url);
      if (status !== null && status >= 400) {
        ctx.brokenLinks.push({ url, status_code: status });
      }
    }
  }

  progress(66, "Analyzing pages");
  return ctx;
}

function decodeText(page: FetchedPage): string {
  const contentType = page.headers["content-type"] || "";
  let charset = "";
  if (contentType.toLowerCase().includes("charset=")) {
    charset = contentType.toLowerCase().split("charset=")[1].split(";")[0].trim().replace(/^"|"$/g, "");
  }
  try {
    return new TextDecoder(charset || "utf-8", { fatal: false }).decode(page.content);
  } catch {
    return new TextDecoder("utf-8", { fatal: false }).decode(page.content);
  }
}
